### PR TITLE
feat: replace bulk enrollment  enrollmentApiClient usage with the new `lms_enroll_user_in_course` function ENT-4746

### PR DESCRIPTION
**Pull request by binodpant**
This newly added function (see https://github.com/edx/edx-platform/pull/28202 ) should help address REST throttling limits hit during bulk enrollment

This PR replaces the call to enrollmentapiClient call in the `enroll_user` function, used within the bulk enrollment related `enroll_licensed_users_in_course` function

It also uses an alternative new function instead of reusing the `enroll_user` function because doing so  requires a lot of changes to the Manage learners use case tests. I decided it's best not to risk doing that in this PR.

But we definitely need to consolidate these `enroll_` methods and functions into better more coherent classes rather than a massive collection of functions

JIRA: ENT-4746

### Manual testing notes:

After enrolling 3 courses x 25 learners locally I got 

```
edx.devstack.lms                  | 2021-07-23 13:58:27,226 INFO 9114 [enterprise.api.v1.views] [user 17] [ip 172.18.0.16] views.py:314 - Successfully bulk enrolled learners: {<User: ecommerce_worker>, <User: login_service_user>, <User: edx>, <User: enterprise_worker>, <User: honor>, <User: audit>, <User: staff>, <User: retirement_service_worker>, <User: discovery_worker>, <User: credentials_worker>, <User: edx_notes_api_worker>, <User: enterprise_learner_test-enterprise>, <User: enterprise_admin_test-enterprise>, <User: enterprise_admin>, <User: enterprise_openedx_operator>, <User: license_manager_worker>, <User: enterprise_catalog_worker>, <User: test-enterprise_learner_10>, <User: test-enterprise_learner_11>, <User: test-enterprise_learner_12>, <User: test-enterprise_learner_13>, <User: test-enterprise_learner_14>, <User: test-enterprise_learner_15>, <User: test-enterprise_learner_16>, <User: test-enterprise_learner_17>} into course course-v1:edX+303707808+2T2021
edx.devstack.lms                  | 2021-07-23 13:58:28,283 INFO 9114 [enterprise.api.v1.views] [user 17] [ip 172.18.0.16] views.py:314 - Successfully bulk enrolled learners: {<User: ecommerce_worker>, <User: login_service_user>, <User: edx>, <User: enterprise_worker>, <User: honor>, <User: audit>, <User: staff>, <User: retirement_service_worker>, <User: discovery_worker>, <User: credentials_worker>, <User: edx_notes_api_worker>, <User: enterprise_learner_test-enterprise>, <User: enterprise_admin_test-enterprise>, <User: enterprise_admin>, <User: enterprise_openedx_operator>, <User: license_manager_worker>, <User: enterprise_catalog_worker>, <User: test-enterprise_learner_10>, <User: test-enterprise_learner_11>, <User: test-enterprise_learner_12>, <User: test-enterprise_learner_13>, <User: test-enterprise_learner_14>, <User: test-enterprise_learner_15>, <User: test-enterprise_learner_16>, <User: test-enterprise_learner_17>} into course course-v1:edX+808404707+2T2021
edx.devstack.lms                  | 2021-07-23 13:58:29,584 INFO 9114 [enterprise.api.v1.views] [user 17] [ip 172.18.0.16] views.py:314 - Successfully bulk enrolled learners: {<User: ecommerce_worker>, <User: login_service_user>, <User: edx>, <User: enterprise_worker>, <User: honor>, <User: audit>, <User: staff>, <User: retirement_service_worker>, <User: discovery_worker>, <User: credentials_worker>, <User: edx_notes_api_worker>, <User: enterprise_learner_test-enterprise>, <User: enterprise_admin_test-enterprise>, <User: enterprise_admin>, <User: enterprise_openedx_operator>, <User: license_manager_worker>, <User: enterprise_catalog_worker>, <User: test-enterprise_learner_10>, <User: test-enterprise_learner_11>, <User: test-enterprise_learner_12>, <User: test-enterprise_learner_13>, <User: test-enterprise_learner_14>, <User: test-enterprise_learner_15>, <User: test-enterprise_learner_16>, <User: test-enterprise_learner_17>} into course course-v1:edX+909878+2T2021
```
**Merge checklist:**
- [ ] Any new requirements are in the right place (do **not** manually modify the `requirements/*.txt` files)
    - `base.in` if needed in production but edx-platform doesn't install it
    - `test-master.in` if edx-platform pins it, with a matching version
    - `make upgrade && make requirements` have been run to regenerate requirements
- [ ] `make static` has been run to update webpack bundling if any static content was updated
- [ ] `./manage.py makemigrations` has been run
    - Checkout the [Database Migration](https://openedx.atlassian.net/wiki/spaces/AC/pages/23003228/Everything+About+Database+Migrations) Confluence page for helpful tips on creating migrations.
    - *Note*: This **must** be run if you modified any models.
      - It may or may not make a migration depending on exactly what you modified, but it should still be run.
    - This should be run from either a venv with all the lms/edx-enterprise requirements installed or if you checked out edx-enterprise into the src directory used by lms, you can run this command through an lms shell.
        - It would be `./manage.py lms makemigrations` in the shell.
- [x] [Version](https://github.com/edx/edx-enterprise/blob/master/enterprise/__init__.py) bumped
- [ ] [Changelog](https://github.com/edx/edx-enterprise/blob/master/CHANGELOG.rst) record added
- [ ] Translations updated (see docs/internationalization.rst but also this isn't blocking for merge atm)

**Post merge:**
- [ ] Tag pushed and a new [version](https://github.com/edx/edx-enterprise/releases) released
    - *Note*: Assets will be added automatically. You just need to provide a tag (should match your version number) and title and description.
- [ ] After versioned build finishes in [GitHub Actions](https://github.com/edx/edx-enterprise/actions), verify version has been pushed to [PyPI](https://pypi.org/project/edx-enterprise/)
    - Each step in the release build has a condition flag that checks if the rest of the steps are done and if so will deploy to PyPi.
    (so basically once your build finishes, after maybe a minute you should see the new version in PyPi automatically (on refresh))
- [ ] PR created in [edx-platform](https://github.com/edx/edx-platform) to upgrade dependencies (including edx-enterprise)
    - This **must** be done after the version is visible in PyPi as `make upgrade` in edx-platform will look for the latest version in PyPi.
    - Note: the edx-enterprise constraint in edx-platform **must** also be bumped to the latest version in PyPi.

### CODE REVIEW

File: enterprise/api/v1/views.py
Comment (binodpant) [inline enterprise/api/v1/views.py:336]: removing now unused function

File: enterprise/utils.py
Comment (binodpant) [inline enterprise/utils.py:1429]: this function is really only used to enroll in one course, and the succeeded flag also becomes clearer in this single course mode hence I changed it

File: enterprise/api/v1/views.py
Comment (binodpant) [inline enterprise/api/v1/views.py:318]: removing this, since a customer-admin-enrollment tracking is already sent from inside the function `customer_admin_enroll_user_in_course` @taliaedX lmk if any flags

File: enterprise/utils.py
Comment (binodpant) [inline enterprise/utils.py:1567]: not refactoring the `enroll_user` function in this PR since I find that Manage learners also uses that function and it is going to risk adding regressions there. I want to remove the use of enroll_user in favor of this, but I also don't want to delay bulk enrollment. Also a lot of tests are written for manage learners that need updates. I would rather do that in a new PR. And we have many places to replace the enrollmentApiClient anyway

**Comment by binodpant** [comment]
I just need to sort out this last test failure, it beats me so far, idk why it only fails for the case of multi course multi learners when checking track_enrollment call. It is making the call from within the function now instead of outside. but I already rewired the mock so ... should have worked..it does work for the other cases just not this last ddt case.

File: tests/test_enterprise/api/test_views.py
Comment (taliaedX) [inline tests/test_enterprise/api/test_views.py:3428]: I think this change right here is why your test is failing @binodpant

File: enterprise/api/v1/views.py
Comment (binodpant) [reply enterprise/api/v1/views.py:318]: undone! we want this tracking too, fixed test as well

File: tests/test_enterprise/api/test_views.py
Comment (binodpant) [reply tests/test_enterprise/api/test_views.py]: yes thanks this was def an issue too

**Comment by binodpant** [comment]
Will do version/changelog after rebase before merge

File: test_utils/fake_enrollment_api.py
Comment (taliaedX) [inline test_utils/fake_enrollment_api.py:167]: this may be a very stupid question, but where/how are we using this fake enroll method?
I don't see anything importing from `fake_enrollment_api` nor even `test_utils`

**Review by taliaedX — COMMENTED** [review]
I have some questions, but overall looks good

File: enterprise/utils.py
Comment (binodpant) [reply enterprise/utils.py:1559]: sorry may be I did not see the issue but looking into a bit... this function is called by two views as well. `create_enterprise_course_enrollment` and `HandleConsentEnrollment#get()`. However it's called not from within `customer_admin_enroll_user` but separately like this in bulk enroll

```
succeeded = customer_admin_enroll_user(enterprise_customer, user, course_mode, course_run_key)
                if succeeded:
                    enterprise_customer_user = get_enterprise_customer_user(user.id, enterprise_customer.uuid)
                    get_create_ent_enrollment(course_run_key, enterprise_customer_user, license_uuid=license_uuid)
                    results['successes'].append({'user': user, 'email': user_email, 'course_run_key': course_run_key})
                else:
```

Comment (binodpant) [reply enterprise/utils.py:1559]: also tracking calls wrt bulk enroll are going to be of this form now ...

`track_enrollment(PATHWAY_CUSTOMER_ADMIN_ENROLLMENT, request.user.id, course_run)`

and are contained in two places

- `customer_admin_enroll_user` in utils.py : bulk enroll specific (for now unless we consolidate/refactor)
- `enroll_learners_in_courses`: also bulk enrollment specific

File: test_utils/fake_enrollment_api.py
Comment (binodpant) [reply test_utils/fake_enrollment_api.py:167]: haha good q let me check if I do actually use it still. I was using it once but maybe left this behind after a nother refactor.. brb

Comment (binodpant) [reply test_utils/fake_enrollment_api.py]: yeah thanks for the catch I removed after verifying it's no longer used

File: enterprise/utils.py
Comment (taliaedX) [reply enterprise/utils.py:1559]: oh yeah I am a dummy. Thank you for verifying! This is what I get for trying to look at the references in PR form.

**Review by taliaedX — APPROVED** [review]
LGTM!

**Comment by binodpant** [comment]
will merge post-lunch after changelog unless @iloveagent57 has any flags/thoughts

File: enterprise/utils.py
Comment (edx-abolger) [reply enterprise/utils.py:1567]: @binodpant I think this is fine, when you make the new ticket can you link it here?

**Pull request merged**
